chain/epoch-manager/src/shard_tracker.rs

```diff
@@ -338,6 +359,68 @@ impl ShardTracker {
             Ok(Some(state_sync_info))
         }
     }
+
+    /// Checks whether `shard_id` is a descendant of any of the `tracked_shards`.
+    /// Assumes that `shard_id` exists in the shard layout of `epoch_id`.
+    fn check_if_descendant_of_tracked_shard(
+        &self,
+        shard_id: ShardId,
+        tracked_shards: &Vec<ShardId>,
+        epoch_id: &EpochId,
+    ) -> Result<bool, EpochError> {
+        if let Some(is_tracked) =
+            self.descendant_of_tracked_shard_cache.lock().unwrap().get(&shard_id)
+        {
+            return Ok(*is_tracked);
+        }
+
+        let is_tracked = check_if_descendant_of_tracked_shard(
+            shard_id,
+            &tracked_shards,
+            &epoch_id,
+            &self.epoch_manager,
+        )?;
+        self.descendant_of_tracked_shard_cache.lock().unwrap().insert(shard_id, is_tracked);
+        Ok(is_tracked)
+    }
+}
+
+fn check_if_descendant_of_tracked_shard(
+    mut shard_id: ShardId,
+    tracked_shards: &Vec<ShardId>,
+    epoch_id: &EpochId,
+    epoch_manager: &Arc<dyn EpochManagerAdapter>,
+) -> Result<bool, EpochError> {
+    if tracked_shards.contains(&shard_id) {
+        // `shard_id` is one of the `tracked_shards` (a shard is a descendant of itself)
+        return Ok(true);
+    }
+    let mut tracked_shards: HashSet<ShardId> = tracked_shards.into_iter().cloned().collect();
+    let mut protocol_version = epoch_manager.get_epoch_protocol_version(epoch_id)?;
+    while protocol_version > PROD_GENESIS_PROTOCOL_VERSION {
+        let shard_layout = epoch_manager.get_shard_layout_from_protocol_version(protocol_version);
+        // Invariants:
+        // * `shard_id` is an ancestor of the original `shard_id` (or itself).
+        // * `shard_id` belongs to the `shard_layout`.
+        // * `tracked_shards` do not contain `shard_id`.
+
+        // `shard_id` cannot be a descendant of another shard from the current shard layout.
+        // It could potentially be a descendant of an earlier shard with the same ID as a shard in the current shard layout.
+        // In that case, we are only interested in the latest occurrence of a shard with the given ID.
+        // To ensure that, we remove the shard IDs of the current shard layout from `tracked_shards`.
+        shard_layout.shard_ids().for_each(|not_tracked_shard| {
+            tracked_shards.remove(&not_tracked_shard);
+        });
+        protocol_version -= 1;
+        let Some(parent_shard_id) = shard_layout.try_get_parent_shard_id(shard_id)? else {
+            continue;
+        };
+        if tracked_shards.contains(&parent_shard_id) {
+            return Ok(true);
+        }
+        shard_id = parent_shard_id;
```

Comment:
I don't think the invariants hold here or at least it's hard to tell. You always go to the parent shard id but for a lot of protocol versions there isn't a shard layout change. It seems like you may switch to the parent shard id without switching to the parent shard layout.

Reply:
changed implementation